Add @woocommerce/create-product-editor-block package

mattsherman opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR introduces a new `@woocommerce/create-product-editor-block` package, which is a template that can be used with the `@wordpress/create-block` tool to quickly get the skeleton of a new block-based extension to the new block-based product editor up-and-running.

The skeleton project generated has support for:

- `wp-env`
    - WordPress 6.2
    - WooCommerce trunk nightly
    - WooCommerce Beta Tester
- linting and code formatting
- TypeScript

The example block generated simply renders a message. Support for interactivity in the example block will be added in future PRs.

<img width="1276" alt="Screenshot 2023-05-15 at 16 43 57" src="https://github.com/woocommerce/woocommerce/assets/2098816/f0dcefbe-d663-424c-b34e-87701b25fd54">


The aim is to get devs up and running with block-based extensibility in a few simple commands.

Closes #38143.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Follow the instructions in the package's README.md (note, since this package hasn't been published yet, you will need to use the `npx @wordpress/create-block --template ./path/to/woocommerce/packages/js/create-product-editor-block my-extension-name` form of the command, with a local path to the template)
2. Verify that the generated skeleton project can be used, with the built-in `wp-env` support, to render the example block in the block-based product form, with no modifications of any files or any commands other than those mentioned in the README.md.
3. Verify the generated skeleton project's files look reasonable.

<!-- End testing instructions -->

mattsherman committed:
Initial commit for create-product-editor-block

mattsherman committed:
Changelog

mattsherman committed:
Add stylelint support

mattsherman committed:
Install WooCommerce and WooCommerce Beta Tester in wp-env

mattsherman committed:
Set PHP and WordPress versions

mattsherman committed:
Enable product block editor after wp-env setup

mattsherman committed:
Lower priority for modifying template, so that the template is actual there!

mattsherman committed:
Update wp-env afterSetup command to use npx

mattsherman committed:
Update output README.md

mattsherman committed:
Update README for package

mattsherman committed:
Use long form of template parm name in README

mattsherman committed:
Remove unused example in block config

mattsherman committed:
Update block registration to handle both WP 6.2 and WP 6.3

mattsherman committed:
Include note about how to change version of WP and WC in README

mattsherman committed:
Update WP to 6.2.1

mattsherman commented:
Pulling this back to `In Progress` to improve the block registration based on Slack discussions. I'll also rebase to pull in the fix for the unrelated failing e2e test.

mattsherman committed:
Remove explicit enqueuing of block script and style

mattsherman commented on packages/js/create-product-editor-block/index.js:
```diff
@@ -0,0 +1,50 @@
+		editorStyle: 'file:./index.css',
+		style: 'file:./index.css',
+		customBlockJSON: {
+			script: 'file:./index.js',
```
I need to include the `script` prop in my `block.json` in order for the block to load in the product editor

louwie17 replied:
Similar to what we discussed in Slack with the latest WP release, I think this is a bug in our initialization code, where only blocks with the `script` are being enqueued.
We can leave this in for now, but I will create a follow up issue/bug to address this.

mattsherman replied:
Issue to address this in product editor initialization: https://github.com/woocommerce/woocommerce/issues/38394

mattsherman replied:
Issue to update the `@woocommerce/create-product-editor-block` package: https://github.com/woocommerce/woocommerce/issues/38426

louwie17 commented on packages/js/create-product-editor-block/plugin-templates/$slug.php.mustache:
```diff
@@ -0,0 +1,85 @@
+{{/description}}
+ * Version:           {{version}}
+ * Requires at least: 6.1
+ * Requires PHP:      7.0
```
Maybe update this to `7.3` to match WooCommerce.

mattsherman replied:
Updated in https://github.com/woocommerce/woocommerce/pull/38263/commits/5965be42582ea3ab2089d822fb2f1ea716da0ef7

I also added a minimum WC version in https://github.com/woocommerce/woocommerce/pull/38263/commits/4e8651e8bf2ef56aee26c727cc974d5da5e8f8b1

louwie17 commented on packages/js/create-product-editor-block/plugin-templates/.wp-env.override.json.mustache:
```diff
@@ -0,0 +1,18 @@
+		"SCRIPT_DEBUG": true,
+		"ALTERNATE_WP_CRON": true,
+	},
+	"afterSetup": "npx wp-env run cli \"wp option update wc_admin_helper_feature_values '{\\\"product-block-editor\\\": true}' --format=json\""
```
This will have to be updated, we recently pushed to PR to re-enable the feature flag, meaning you have to set `woocommerce_feature_product_block_editor_enabled` option to `yes`:
`wp option update woocommerce_feature_product_block_editor_enabled 'yes'`

mattsherman replied:
Added in https://github.com/woocommerce/woocommerce/pull/38263/commits/a1acad696b926b266167d5733c3c8dcdcdfc48e5

louwie17 commented on packages/js/create-product-editor-block/plugin-templates/.wp-env.override.json.mustache:
```diff
@@ -0,0 +1,18 @@
+	"core": "https://wordpress.org/wordpress-6.2.1.zip",
+	"plugins": [
+		"https://github.com/woocommerce/woocommerce/releases/download/nightly/woocommerce-trunk-nightly.zip",
+		"https://github.com/woocommerce/woocommerce/releases/download/wc-beta-tester-2.3.0/woocommerce-beta-tester.zip",
```
Related to the below comment, this means the beta tester is not required anymore.

mattsherman replied:
While not required, it seems like it could still be a good idea to include, since it exposes 3PDs to the beta tester, which can be a helpful tool. What do you think?

louwie17 replied:
Yeah I think that makes sense, sounds good with me :)

louwie17 reviewed (commented):
Nice work @mattsherman 🎉 This tested well, I did leave a couple minor suggestions before merging this in, but otherwise this looks great!

mattsherman committed:
Update minimum required versions of WordPress and PHP in generated plugin file

mattsherman committed:
Enable feature flag option after setup of wp-env

mattsherman committed:
Add minimum required WC version to generated plugin file

mattsherman commented:
@louwie17 Thanks for the review. I've addressed all of your feedback (leaving the beta tester plugin in for now, unless you want it removed). Ready for re-review.

louwie17 commented on packages/js/create-product-editor-block/plugin-templates/.wp-env.override.json.mustache:
```diff
@@ -0,0 +1,18 @@
+		"SCRIPT_DEBUG": true,
+		"ALTERNATE_WP_CRON": true,
+	},
+	"afterSetup": "npx wp-env run cli \"wp option update woocommerce_feature_product_block_editor_enabled 'yes'\""
```
So it looks like wpEnv just deprecated this config in replacement of `lifecycleScripts` -> https://github.com/WordPress/gutenberg/blob/trunk/packages/env/CHANGELOG.md#breaking-change

I believe you can replace it with this:
```
"lifecycleScripts": {
		"afterStart": "npx wp-env run cli \"wp option update woocommerce_feature_product_block_editor_enabled 'yes'\""
	}
```

ObliviousHarmony replied:
We also updated the `wp-env run` syntax so that it operates in the same way that `docker-compose run` does. This means that quoting the entire string is no longer going to get the behavior you want. You would want to replace it with

```
"lifecycleScripts": {
		"afterStart": "npx wp-env run cli wp option update woocommerce_feature_product_block_editor_enabled 'yes'"
}
```

mattsherman replied:
@ObliviousHarmony Thanks -- I saw that in the wp-env changelog!

mattsherman replied:
Done in https://github.com/woocommerce/woocommerce/pull/38263/commits/0bf1ca7ded699dbee1fe9c2939e63791b4089f0c

louwie17 commented on packages/js/create-product-editor-block/plugin-templates/$slug.php.mustache:
```diff
@@ -0,0 +1,86 @@
+add_action( 'init', '{{namespaceSnakeCase}}_{{slugSnakeCase}}_block_init' );
+
+function {{namespaceSnakeCase}}_{{slugSnakeCase}}_add_block_to_product_editor( $args ) {
+	$template = $args['template'];
```
Now that our template addition is only added when the feature flag is enabled, we should just add a check here to see if `template` is defined.
At some point we may want to check for the for the feature flag as well.
So could add this:
```
if ( ! isset( $args['template'] ) ) {
  return $args;
}
```
We could add these conditions:
```
class_exists( '\Automattic\WooCommerce\Utilities\FeaturesUtil' ) && \Automattic\WooCommerce\Utilities\FeaturesUtil::feature_is_enabled('product_block_editor')
```

mattsherman replied:
Done in https://github.com/woocommerce/woocommerce/pull/38263/commits/8f6ba153406d054b15ccba2a10a8c59ae39f625f

louwie17 reviewed (changes requested):
Ran into a couple more small issues, I would of almost pushed these changes up myself and merge it, but thought I wait, they should be relatively minimal :)

mattsherman commented:
> Ran into a couple more small issues, I would of almost pushed these changes up myself and merge it, but thought I wait, they should be relatively minimal :)

Thanks for catching these! I'll make those changes this morning and ping you when ready for another look!

mattsherman committed:
Support wp-env lifecycleScripts

mattsherman committed:
Guard against product_block_editor not enabled or missing template

mattsherman commented:
Ready for re-review @louwie17 !

louwie17 reviewed (dismissed):
LGTM 🚀 Nice work @mattsherman, looking forward to getting this in.
Unfortunately it looks like there is a merge conflict with the pnpm lock file now 😢

mattsherman commented:
> LGTM 🚀 Nice work @mattsherman, looking forward to getting this in. Unfortunately it looks like there is a merge conflict with the pnpm lock file now 😢

I feared that. I'll fix it up and put it back up for re-review/approval.

mattsherman committed:
Update lock file

mattsherman commented:
@louwie17 Ready for re-review... the only change is the lock file update.

louwie17 reviewed (approved):
LGTM 🚀 Thanks for fixing the merge conflict!
